zppy/e3sm_diags.py

```diff
@@ -33,10 +33,12 @@ def e3sm_diags(config, scriptDir, existing_bundles, job_ids_file):  # noqa: C901
         return existing_bundles
 
     # --- Generate and submit e3sm_diags scripts ---
-    dependencies: List[str] = []
```

Comment:
@forsyth2 Thank for looking into this. To start fresh dependencies list for each `e3sm_diags` sub tasks, we can just reset `dependencies = []`, when looping over tasks, i.e. under line 38 in e3sm_diags.py without touching other part of zppy.  I'm not sure why it is needed to introduce another parameter called `carried_over_dependencies`?

Reply:
@chengzhuzhang we can't start a completely fresh dependencies list on each new task because there is a case where we do want to "remember" dependencies from the previous one. This happens when `tc_analysis` is a dependency. 

In https://github.com/E3SM-Project/zppy/blob/main/zppy/e3sm_diags.py:
```
                    # Due to a `socket.gaierror: [Errno -2] Name or service not known` error when running e3sm_diags with tc_analysis
                    # on multiple year_sets, if tc_analysis is in sets, then e3sm_diags should be run sequentially.
                    if "tc_analysis" in c["sets"]:
                        # Note that this line should still be executed even if jobid == -1
                        # The later tc_analysis-using e3sm_diags tasks still depend on this task (and thus will also fail).
                        # Add to the dependency list
                        dependencies.append(statusFile)
```

A socket error occurs if we try to run `e3sm_diags` in parallel on `tc_analysis`. To work around that, we simply make `e3sm_diags` run sequentially when it's working with `tc_analysis`. (`dependencies.append(statusFile)` adds the current task to the dependencies for the next task -- _this_ is the point of `carried_over_dependencies`). This is similar to how we make `mpas_analysis` run sequentially.

The unfortunate side effect is that `e3sm_diags` tasks that do _not_ deal with `tc_analysis` also end up caught in the sequential dependency chaining. And that means they don't run if an earlier task fails.

Reply:
I'm curious how `tc_analysis` dependency works differently than other dependencies such as climo and time-series. I'm interested in reproducing the `socket error`. `To try to run e3sm_diags in parallel on tc_analysis`, does it mean, when tc_analysis task is on, while `tc_analysis` set is not set to be one of `e3sm_diags` sets to run, so that `tc_analysis` and `e3sm-diags` run in parallel?

Reply:
> I'm curious how `tc_analysis` dependency works differently than other dependencies such as climo and time-series.

In this pull request, it's this line in the task we're finishing up
```
carried_over_dependencies.append(statusFile)
```
and this line in the next task:
```
dependencies: List[str] = carried_over_dependencies
```
This resets all other dependencies, except for the one we specifically added to `carried_over_dependencies` (i.e., the last task).

>  I'm interested in reproducing the `socket error`

This error occurs intermittently when running two `e3sm_diags` jobs, both of which run the `tc_analysis` set. The sequential workaround was introduced in #169. A full discussion of the `tc_analysis` concurrency/parallelism bugs can be found in #180.